pkg/mads/v1/reconcile/snapshot_generator.go

```diff
@@ -26,34 +35,97 @@ type snapshotGenerator struct {
+	meshes, err := s.getMeshesWithPrometheusEnabled(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	meshGateways, err := s.getMeshGateways(ctx, meshes)
-	if err != nil {
-		return nil, err
+	if len(meshes) > 0 && len(meshMetrics) > 0 {
+		log.Info("it is not supported to use both MeshMetrics policy and 'metrics' under Mesh resource. If migrating please remove the 'metrics' section and apply an equivalent MeshMetrics resource")
 	}
 
-	args := generator.Args{
-		Meshes:       meshes,
-		Dataplanes:   dataplanes,
-		MeshGateways: meshGateways,
+	var resources []*core_xds.Resource
+	if len(meshMetrics) == 0 {
+		dataplanes, err := s.getDataplanes(ctx, meshes)
+		if err != nil {
+			return nil, err
+		}
+
+		meshGateways, err := s.getMeshGateways(ctx, meshes)
+		if err != nil {
+			return nil, err
+		}
+
+		args := generator.Args{
+			Meshes:       meshes,
+			Dataplanes:   dataplanes,
+			MeshGateways: meshGateways,
+		}
+
+		resources, err = s.resourceGenerator.Generate(args)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		dataplanes, err := s.getMatchingDataplanes(ctx, meshMetrics)
+		if err != nil {
+			return nil, err
+		}
+
+		resources, err = meshmetrics.Generate(dataplanes)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	resources, err := s.resourceGenerator.Generate(args)
-	if err != nil {
+	return mads_v1_cache.NewSnapshot("", core_xds.ResourceList(resources).ToIndex()), nil
+}
+
+func (s *snapshotGenerator) getMeshMetricsWithCorrespondingPrometheusBackend(ctx context.Context, clientId string) ([]*v1alpha1.MeshMetricResource, error) {
+	meshMetricsList := v1alpha1.MeshMetricResourceList{}
+	if err := s.resourceManager.List(ctx, &meshMetricsList); err != nil {
 		return nil, err
 	}
 
-	return mads_v1_cache.NewSnapshot("", core_xds.ResourceList(resources).ToIndex()), nil
+	meshMetrics := make([]*v1alpha1.MeshMetricResource, 0)
+	for _, meshMetric := range meshMetricsList.Items {
+		for _, backend := range *meshMetric.Spec.Default.Backends { // can backends be nil?
+			// match against client ID or fallback to "" when specified by user
+			if backend.Type == v1alpha1.PrometheusBackendType && (*backend.Prometheus.ClientId == clientId || *backend.Prometheus.ClientId == "") {
+				meshMetrics = append(meshMetrics, meshMetric)
+			}
+		}
+	}
+
+	return meshMetrics, nil
+}
+
+func (s *snapshotGenerator) getMatchingDataplanes(ctx context.Context, meshMetrics []*v1alpha1.MeshMetricResource) (map[*v1alpha1.MeshMetricResource]*core_mesh.DataplaneResourceList, error) {
+	meshMetricToDataplanes := map[*v1alpha1.MeshMetricResource]*core_mesh.DataplaneResourceList{}
+
+	for _, meshMetric := range meshMetrics {
+		dataplaneList := &core_mesh.DataplaneResourceList{}
+		filter := func(rs model.Resource) bool {
+			dpp := rs.(*core_mesh.DataplaneResource)
+			res, _ := matchers.PolicyMatches(meshMetric, dpp, xds_context.NewResources()) // what are the dependant resources? do I need them?
+			return res
+		}
```

Comment:
This is similar to what we filter in `insights` we get all dpps and filter by the ones that match the policy. I guess in insights it might be ok because it's on demand. Here we will be doing this on interval (I think it's 15s for prometheus).